add sym_strides() function, use in fake/proxy tensors

bdhirsh opened the pull request:
Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81302
* #81301
* #81238
* __->__ #81300

Differential Revision: [D38594558](https://our.internmc.facebook.com/intern/diff/D38594558)

bdhirsh committed:
add sym_strides() function, use in fake/proxy tensors

[ghstack-poisoned]

bdhirsh commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -187,7 +189,13 @@ def __new__(cls, elem, proxy, *, requires_grad=None):
+            ], dtype=elem.dtype, layout=elem.layout, device=elem.device, requires_grad=elem.requires_grad,
+            strides=[
+                create_proxy_symint(stride_i, proxy.stride(i)) for i, stride_i in enumerate(create_contiguous(elem.shape))
+            ], storage_offset=elem.storage_offset())
```
@Chillee I wonder if this is a footgun. Before, the symbolic sizes code here had a proxy associated with it but the strides didn't. So in C++ if you try to do something like `mul(size[i], stride[i])`, I ran into trouble because the two backing objects are now `ProxySymInt` and `PySymInt`, which ends up surfacing as an error inside of `fx.Proxy`.

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"




[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"




[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"




[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/NestedTensorImpl.cpp:
```diff
@@ -153,6 +153,13 @@ c10::SymIntArrayRef NestedTensorImpl::sym_sizes_custom() const {
+}
+
+c10::SymIntArrayRef NestedTensorImpl::sym_strides() const {
+  return sym_strides_custom();
```
It's weird that we have to implement these...

bdhirsh replied:
whoops this was a mistake

bdhirsh replied:
whoops, this was a mistake

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2430,4 +2434,46 @@ c10::SymIntArrayRef concrete_sym_sizes_fn(
+  c10::SymInt* start = (c10::SymInt*)result[0];
+  int64_t len = result[1];
+
+  return c10::SymIntArrayRef(start, len);
```
It would be nice if the epilogue here was deduplciated

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.






[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38594558).

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38594558).

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.




Differential Revision: [D38594558](https://our.internmc.facebook.com/intern/diff/D38594558)

[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.




Differential Revision: [D38594558](https://our.internmc.facebook.com/intern/diff/D38594558)

[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.




Differential Revision: [D38594558](https://our.internmc.facebook.com/intern/diff/D38594558)

[ghstack-poisoned]

bdhirsh committed:
Update on "add sym_strides() function, use in fake/proxy tensors"

Add `TensorImpl::sym_strides`, bind it to python with `torch.ops.aten.sym_strides`, and use it in `ProxyTensor` and `FakeTensor`.

Before, `ProxyTensor` was generating `ProxySymInt`'s for the sizes, but not for the strides. Internally we still represent strides with a `SymIntArrayRef` though, so I ran into some weird issues where sizes were showing up as `ProxySymInt`, but strides were `PySymInt`'s.




Differential Revision: [D38594558](https://our.internmc.facebook.com/intern/diff/D38594558)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
